### DISCUSSION TITLE
A case with data query

### DISCUSSION THREAD

**Discussion by roman-mazur** [cue-lang/cue#2616] [Q&A] [answered]
Hello,

I'm experimenting with the following sample where I query some data and then use the query result in some transformation.

https://cuelang.org/play/?id=6qSfDcl8A2m#cue@export@cue

```
#Data: {x: string, y: string}

All: [...#Data]
All: [{x: "a", y: "b"}, {x: "c", y: "d"}]

#Get: {
  #filter: string
  
  _#search: [for d in All if d.x == #filter {d}]
  _#valid: len(_#search) & 1
  
  _#search[0]
}

query: #Get & { #filter: "a" }

#Stringify: {
  #input: #Data
  "x=\(#input.x)  y=\(#input.y)"
}

str: {#Stringify, #input: query}
```

From my point of view this should work, but I'm getting an error. Try commenting out the line with `str:` - you will see that the query is successful, it resolves to a concrete value. However, when I refer to this result from a different place, it fails with a mysterious error.

I would appreciate somebody's look to understand if it should be submitted as a bug.

**Comment by roman-mazur** [comment]
If I change `#Get` to not use the embedding, but expose a separate field, it works.
```
#Get: {
  #filter: string
  
  _#search: [for d in All if d.x == #filter {d}]
  _#valid: len(_#search) & 1
  
  result: _#search[0]
}

query: (#Get & { #filter: "a" }).result
```
https://cuelang.org/play/?id=Go4fW9WFF2m#cue@export@cue

**Reply by verdverm** [reply]
Another variation: https://cuelang.org/play?id=59Aj4v5cBex

```cue
// from: https://cuelang.org/play/?id=6qSfDcl8A2m

#Data: {x: string, y: string}

All: [...#Data]
All: [{x: "a", y: "b"}, {x: "c", y: "d"}]

#Get: {
  #filter: string
  
  _#search: [for d in All if d.x == #filter {d}]
  _#valid: len(_#search) & 1
  
  _#search[0]
}


#Stringify: {
  #input: #Data
  s: "x=\(#input.x)  y=\(#input.y)"
}

str: #Stringify & {#input: #Get & { #filter: "a" }}
```

If I turn `#Stringify.s` into an embedded string, the same error comes back

**Comment by verdverm** [comment]
This looks like a closedness issue, related to embedding and embedded scalars

There are a number of open issues / questions around embedding, especially as it comes to definitions and closedness

- https://github.com/cue-lang/cue/milestone/17  (multiple)
- https://github.com/cue-lang/cue/issues/543 (proposal)

**Reply by roman-mazur** [reply, marked as answer]
Marking this as answered: the original example does work with the latest CUE release.
